# Python Basic

**1. Write a Python program to print the following string in a specific format (see the output).**<br>

<i>Sample String:</i> 

"Twinkle, twinkle, little star, How I wonder what you are! Up above the world so high, Like a diamond in the sky. Twinkle, twinkle, little star, How I wonder what you are"<br>

<i>Output:</i>
<pre>
Twinkle, twinkle, little star,
	How I wonder what you are!
		Up above the world so high,		
		Like a diamond in the sky.
Twinkle, twinkle, little star,
	How I wonder what you are</pre>

In [1]:
# `\n` symbol means literally new line
# `\t` symbol means add a tab
print("Twinkle, twinkle, little star, \n\tHow I wonder what you are! \n\t\tUp above the world so high, \n\t\tLike a diamond in the sky. \nTwinkle, twinkle, little star, \n\tHow I wonder what you are")

Twinkle, twinkle, little star, 
	How I wonder what you are! 
		Up above the world so high, 
		Like a diamond in the sky. 
Twinkle, twinkle, little star, 
	How I wonder what you are


**2. Write a Python program to find out what version of Python you are using.**

In [2]:
# [1] `sys` module
import sys

print("Python Version")
print(sys.version)
print("Version Info")
print(sys.version_info)

Python Version
3.8.8 (default, Apr 13 2021, 19:58:26) 
[GCC 7.3.0]
Version Info
sys.version_info(major=3, minor=8, micro=8, releaselevel='final', serial=0)


In [3]:
# [2] Run shell command in Jupyter Notebook
!python -V

Python 3.8.8


In [4]:
# [3] Run shell command in Jupyter Notebook
!python --version

Python 3.8.8


In [5]:
# [4] `platform` module
import platform

print(platform.python_version())

3.8.8


**3. Write a Python program to display the current date and time.**<br>

<i>Sample Output:</i>

Current date and time:<br>
2014-07-05 14:34:14

In [6]:
# `datetime.datetime.now()` & `date.strftime(format)`
import datetime

now = datetime.datetime.now()
print("Current date and time:")
print(now.strftime("%Y-%m-%d %H:%M:%S"))

Current date and time:
2023-05-17 12:13:52


**4. Write a Python program that calculates the area of a circle based on the radius entered by the user.**

<i>Sample Output:</i><br>

r = 1.1<br>
Area = 3.8013271108436504

In [14]:
# `math.pi` & `input()`
from math import pi

r = float(input("Input the radius of the circle: "))

print("The area of the circle with radius " + str(r) + " is: " + str(pi * r**2))

Input the radius of the circle: 10
The area of the circle with radius 10.0 is: 314.1592653589793


**5. Write a Python program that accepts the user's first and last name and prints them in reverse order with a space between them.**

In [8]:
# `input()`
fname = input("Input your First Name: ")
lname = input("Input your Last Name: ")

print("Hello " + lname + " " + fname)

Input your First Name: Yung-Shun
Input your Last Name: Chang
Hello Chang Yung-Shun


**6. Write a Python program that accepts a sequence of comma-separated numbers from the user and generates a list and a tuple of those numbers.**

<i>Sample Data:</i><br>
<br>
3, 5, 7, 23

<i>Output:</i><br>

List : ['3', ' 5', ' 7', ' 23']<br>
Tuple : ('3', ' 5', ' 7', ' 23')

In [9]:
# `split()` for a list & `tuple()` for a tuple
values = input("Input some comma-separated numbers: ")

list = values.split(",")
tuple = tuple(list)

print("List: ", list)
print("Tuple: ", tuple)

Input some comma-separated numbers: 3,5,8,17
List:  ['3', '5', '8', '17']
Tuple:  ('3', '5', '8', '17')


**7. Write a Python program that accepts a filename from the user and prints the extension of the file.**

<i>Sample Filename:</i><br>

abc.java

<i>Output:</i>

java

In [10]:
# `split()` & `-1` as index
filename = "abc.java"
f_extns = filename.split(".")
print("The extension of the file is: " + repr(f_extns[-1]))

The extension of the file is: 'java'


**8. Write a Python program to display the first and last colors from the following list.**

color_list = ["Red","Green","White" ,"Black"]

In [11]:
# `%s`
color_list = ["Red", "Green", "White", "Black"]
print("%s %s" % (color_list[0], color_list[-1]))

Red Black


**9. Write a Python program to display the examination schedule. (extract the date from exam_st_date).**

exam_st_date = (11, 12, 2014)

<i>Sample Output:</i>

The examination will start from : 11 / 12 / 2014

In [12]:
# `%i` formats the integers
exam_st_date = (11, 12, 2014)
print("The examination will start from: %i / %i / %i" % exam_st_date)

The examination will start from: 11 / 12 / 2014


**10. Write a Python program that accepts an integer (n) and computes the value of n+nn+nnn.**

<i>Sample Value:</i> 

n = 5

<i>Expected Result:</i> 

615

In [13]:
# `int()` & `%s`
n = int(input("Input an integer: "))

n1 = int("%s" % n)
n2 = int("%s%s" % (n, n))
n3 = int("%s%s%s" % (n, n, n))

print(n1 + n2 + n3)

Input an integer: 5
615


**11. Write a Python program to print the documents (syntax, description etc.) of Python built-in function(s).**

<i>Sample Function:</i> 

abs()

<i>Expected Result:</i>

abs(number) -> number<br>
Return the absolute value of the argument.

In [14]:
# `__doc__`
print(abs.__doc__)

Return the absolute value of the argument.


**12. Write a Python program that prints the calendar for a given month and year.**

<i>Note:</i>

Use "calendar" module.

In [15]:
# `calendar.month()`
import calendar

y = int(input("Input the year: "))
m = int(input("Input the month: "))

print(calendar.month(y, m))

Input the year: 2023
Input the month: 5
      May 2023
Mo Tu We Th Fr Sa Su
 1  2  3  4  5  6  7
 8  9 10 11 12 13 14
15 16 17 18 19 20 21
22 23 24 25 26 27 28
29 30 31



**13. Write a Python program to print the following "here document".**

<i>Sample String:</i>
<pre>
a string that you "don't" have to escape
This
is a ....... multi-line
heredoc string --------> example</pre>

In [16]:
# Here document (heredoc)
# `print("""text""")`
print("""
a string that you "don't" have to escape
This
is a ....... multi-line
heredoc string --------> example
""")


a string that you "don't" have to escape
This
is a ....... multi-line
heredoc string --------> example



**14. Write a Python program to calculate the number of days between two dates.**

<i>Sample Dates:</i> 

(2014, 7, 2), (2014, 7, 11)

<i>Expected Output:</i> 

9 days

In [17]:
# `datetime.date(year, month, day)` & `timedelta.days`
# timedelta = date1 - date2
from datetime import date

f_date = date(2014, 7, 2)
l_date = date(2014, 7, 11)
delta = l_date - f_date
print(delta.days)

9


**15. Write a Python program to get the volume of a sphere with radius six.**

In [18]:
# `math.pi`
from math import pi

r = 6.0
V = 4.0 / 3.0 * pi * r**3
print("The volume of the sphere is:", V)

The volume of the sphere is: 904.7786842338603


**16. Write a Python program to calculate the difference between a given number and 17. If the number is greater than 17, return twice the absolute difference.**

In [19]:
# if-else statement
def difference(n):
    if n <= 17:
        return 17 - n
    else:
        return (n - 17) * 2

print(difference(3))
print(difference(36))

14
38


**17. Write a Python program to test whether a number is within 100 of 1000 or 2000.**

In [20]:
# if-else statement & `or`
def near_thousand(n):
    """
    if abs(n - 1000) <= 100 or abs (n - 2000) <= 100:
        return True
    else:
        return False
    """
    return ((abs(n - 1000) <= 100) or (abs(n - 2000) <= 100))
    
print(near_thousand(1024))
print(near_thousand(2048))
print(near_thousand(3178))

True
True
False


**18. Write a Python program to calculate the sum of three given numbers. If the values are equal, return three times their sum.**

In [21]:
# if-else statement & `x == y == z`
def sum_thrice(x, y, z):
    result = x + y + z  
    if x == y == z:
        result = result * 3
    return result
    
print(sum_thrice(1986, 3, 17))
print(sum_thrice(16, 16, 16))

2006
144


**19. Write a Python program to get a newly-generated string from a given string where "Is" has been added to the front. Return the string unchanged if the given string already begins with "Is".**

In [22]:
# `len()` & string slicing with `:` delimiter
def new_string(text):
    if len(text) >= 2 and text[:2] == "Is":
        return text
    return "Is" + text

print(new_string("Array"))
print(new_string("IsEmpty"))

IsArray
IsEmpty


**20. Write a Python program that returns a string that is n (non-negative integer) copies of a given string.**

In [23]:
# [1] `*` operator
def larger_string(text, n):
    return text * n

print(larger_string("love", 3))

lovelovelove


In [24]:
# [2] `+` operator & `range()`
def larger_string(text, n):
    result = ""
    for i in range(n):
        result = result + text
    return result

print(larger_string("hate", 3))

hatehatehate


**21. Write a Python program that determines whether a given number (accepted from the user) is even or odd, and prints an appropriate message to the user.**

In [25]:
# `%` operator
num = int(input("Enter a number: "))

mod = num % 2
if mod > 0:
    print("This is an odd number.")
else:
    print("This is an even number.")

Enter a number: 317
This is an odd number.


**22. Write a Python program to count the number 4 in a given list.**

In [26]:
# [1] for loop & list iteration
def list_count_4(nums):
    count = 0
    for num in nums:
        if num == 4:
            count = count + 1
    return count

print(list_count_4([1, 2, 1, 2, 4, 5, 6, 7, 8, 9, 0]))
print(list_count_4([5, 4, 1, 7, 5, 4, 3, 0, 1, 0, 4]))

1
3


In [27]:
# [2] while loop & list iteration
def list_count_4(nums):
    count = 0
    i = 0
    while i != len(nums):
        if nums[i] == 4:
            count = count + 1
        i = i + 1
    return count

print(list_count_4([1, 2, 1, 2, 4, 5, 6, 7, 8, 9, 0]))
print(list_count_4([5, 4, 1, 7, 5, 4, 3, 0, 1, 0, 4]))

1
3


**23. Write a Python program to get n (non-negative integer) copies of the first 2 characters of a given string. Return n copies of the whole string if the length is less than 2.**

In [28]:
# [1] `*` operator
def substring_copy(text, n):
    if len(text) < 2:
        return n * text
    else: 
        return n * text[:2]
    
print(substring_copy("hello", 5))
print(substring_copy("a", 5))

hehehehehe
aaaaa


In [29]:
# [2] `+` operator
def substring_copy(text, n):
    flen = 2
    if flen > len(text):
        flen = len(text)
    substr = text[:flen]
    result = ""
    for i in range(n):
        result = result + substr
    return result

print(substring_copy("strings", 3))
print(substring_copy("p", 3))

ststst
ppp


**24. Write a Python program to test whether a passed letter is a vowel or not.**

In [30]:
# [1] `in` & list declaration
def is_vowel(char):
    # if char is lowercase
    vowel = ["a", "e", "i", "o", "u"]
    return char in vowel
    
print(is_vowel("u"))
print(is_vowel("j"))

True
False


In [31]:
# [2] `in` & string declaration
def is_vowel(char):
    # if char is lowercase
    vowel = "aeiou"
    return char in vowel
    
print(is_vowel("e"))
print(is_vowel("q"))

True
False


**25. Write a Python program that checks whether a specified value is contained within a group of values.**

<i>Test Data:</i>

3 -> [1, 5, 8, 3] : True<br>
-1 -> [1, 5, 8, 3] : False

In [32]:
# Python list
def is_group_member(group_data, n):
    """
    for value in group_data:
        if n == value:
            return True
    return False
    """
    return n in group_data

print(is_group_member([3, 5, 8, 16], 5))
print(is_group_member([3, 5, 8, 16], 13))

True
False


**26. Write a Python program to create a histogram from a given list of integers.**

In [33]:
# [1] `*` operator & print directly
def histogram(items):
    for n in items:
        print("*" * n)
        
histogram([1, 9, 8, 6, 3, 1, 7])

*
*********
********
******
***
*
*******


In [34]:
# [2] `+` operator & while loop
def histogram(items):
    for n in items:
        output = ""
        times = n
        while times > 0:
            output = output + "*"
            times = times - 1
        print(output)
        
histogram([6, 1, 5, 3, 7, 9])

******
*
*****
***
*******
*********


**27. Write a Python program that concatenates all elements in a list into a string and returns it.**

In [35]:
# `in` & string concatenation
def concatenate_list_data(list):
    result = ""
    for element in list:
        result = result + str(element)
    return result

print(concatenate_list_data(["Write", " ", "a Python program", " that" + " concatenates " + "all elements", "."]))

Write a Python program that concatenates all elements.


**28. Write a Python program to print all even numbers from a given list of numbers in the same order and stop printing any after 237 in the sequence.**

<i>Sample Numbers List:</i>

<pre>numbers = [    
    386, 462, 47, 418, 907, 344, 236, 375, 823, 566, 597, 978, 328, 615, 953, 345, 
    399, 162, 758, 219, 918, 237, 412, 566, 826, 248, 866, 950, 626, 949, 687, 217, 
    815, 67, 104, 58, 512, 24, 892, 894, 767, 553, 81, 379, 843, 831, 445, 742, 717, 
    958, 743, 527
    ]</pre>

In [36]:
# `break` in loop
numbers = [    
    386, 462, 47, 418, 907, 344, 236, 375, 823, 566, 597, 978, 328, 615, 953, 345, 
    399, 162, 758, 219, 918, 237, 412, 566, 826, 248, 866, 950, 626, 949, 687, 217, 
    815, 67, 104, 58, 512, 24, 892, 894, 767, 553, 81, 379, 843, 831, 445, 742, 717, 
    958, 743, 527
    ]

for item in numbers:
    if item == 237:
        print(item)
        break
    elif item % 2 == 0:
        print(item)

386
462
418
344
236
566
978
328
162
758
918
237


**29. Write a Python program that prints out all colors from color_list_1 that are not present in color_list_2.**

<i>Test Data:</i>

color_list_1 = set(["White", "Black", "Red"])<br>
color_list_2 = set(["Red", "Green"])

<i>Expected Output:</i>

{'Black', 'White'}

In [37]:
# `different()` for sets
color_list_1 = set(["White", "Black", "Red"])
color_list_2 = set(["Red", "Green"])
print("Original set elements:")
print(color_list_1)
print(color_list_2)
print("\nDifferent of color_list_1 and color_list_2:")
print(color_list_1.difference(color_list_2))
print("\nDifferent of color_list_2 and color_list_1:")
print(color_list_2.difference(color_list_1))

Original set elements:
{'White', 'Black', 'Red'}
{'Green', 'Red'}

Different of color_list_1 and color_list_2:
{'White', 'Black'}

Different of color_list_2 and color_list_1:
{'Green'}


**30. Write a Python program that will accept the base and height of a triangle and compute its area.**

In [38]:
# `input`
b = int(input("Input the base: "))
h = int(input("Input the height: "))

area = b * h / 2

print("area = ", area)

Input the base: 4
Input the height: 5
area =  10.0


**31. Write a Python program that computes the greatest common divisor (GCD) of two positive integers.**

In [39]:
# [1] while loop & `min()`
def gcd(x, y):
    i = min(x, y)
    while i != 0:
        if x % i == 0 and y % i == 0:
            return i
        else: 
            i = i - 1

print(gcd(100, 120))
print(gcd(13, 17))
print(gcd(48, 16))

20
1
16


In [40]:
# [2] for loop using `range()` in reverse
def gcd(x, y):
    gcd = 1
    # assuming x >= y
    if x % y == 0:
        return y
    for k in range(int(y/2), 0, -1):
        if x % k == 0 and y % k == 0:
            gcd = k
            break
    return gcd

print(gcd(100, 120))
print(gcd(13, 17))
print(gcd(16, 48))

20
1
16


**32. Write a Python program to find the least common multiple (LCM) of two positive integers.**

In [41]:
# [1] infinite loop & `max()`
def lcm(x, y):
    i = max(x, y)
    while True:
        if i % x == 0 and i % y == 0:
            return i
        i = i + 1
        
print(lcm(13, 23))
print(lcm(20, 160))
print(lcm(16, 36))

299
160
144


In [42]:
# [2] infinite loop & if-else statement
def lcm(x, y):
    if x > y:
        z = x
    else:
        z = y
    while(True):
        if ((z % x == 0) and (z % y == 0)):
            lcm = z
            break
        z = z + 1
    return lcm

print(lcm(13, 23))
print(lcm(20, 160))
print(lcm(36, 16))

299
160
144


**33. Write a Python program to sum three given integers. However, if two values are equal, the sum will be zero.**

In [43]:
# `or`
def sum_three(x, y, z):
    if x == y or y == z or x == z:
        return 0
    else:
        return x + y + z

print(sum_three(3, 5, 7))
print(sum_three(3, 3, 9))

15
0


**34. Write a Python program to sum two given integers. However, if the sum is between 15 and 20 it will return 20.**

In [44]:
# comparison operators
def sum_numbers(x, y):
    result = x + y
    # `<= 20` or `< 20` is not important in this case
    if result >= 15 and result <= 20:
        return 20
    else:
        return result
    
print(sum_numbers(13, 36))
print(sum_numbers(7, 8))

49
20


In [45]:
# `range()`
def sum_numbers_numbers(x, y):
    result = x + y
    if result in range(15, 20):
        return 20
    else:
        return result
    
print(sum_numbers(13, 36))
print(sum_numbers(7, 8))

49
20


**35. Write a Python program that returns true if the two given integer values are equal or their sum or difference is 5.**

In [46]:
# if-else statement & `abs()`
def test_number5(x, y):
    """
    if x == y or abs(x - y) == 5 or x + y == 5:
        return True
    else:
        return False
    """
    return x == y or x + y == 5 or abs(x - y) == 5

print(test_number5(3, 2))
print(test_number5(3, 8))
print(test_number5(5, 5))
print(test_number5(13, 16))
print(test_number5(1, 100))

True
True
True
False
False


**36. Write a Python program to add two objects if both objects are integers.**

In [47]:
# `isinstance()`
def add_numbers(x, y):
    if not (isinstance(x, int) and isinstance(y, int)):
        return "Inputs must be integers!"
    return x + y
    
print(add_numbers(3, 5))
print(add_numbers("Hello", 8))

8
Inputs must be integers!


**37. Write a Python program that displays your name, age, and address on three different lines.**

In [48]:
# `format()`
def personal_details():
    name, age = "Joe", 37
    address = "Taipei, Taiwan"
    print("Name: {}\nAge: {}\nAddress: {}".format(name, age, address))
    
personal_details()

Name: Joe
Age: 37
Address: Taipei, Taiwan


**38. Write a Python program to solve (x + y) * (x + y).**

<i>Test Data:</i>

x = 4, y = 3

<i>Expected Output:</i>

(4 + 3) ^ 2) = 49

In [49]:
# function
def square_of_sum(x, y):
    return (x + y) * (x + y)

print(square_of_sum(3, 5))
print(square_of_sum(5, 8))

64
169


In [50]:
# no function & `format`
x, y = 3, 5
result = x * x + 2 * x * y + y * y
print("({} + {}) ^ 2 = {}".format(x, y, result))

(3 + 5) ^ 2 = 64


**39. Write a Python program to compute the future value of a specified principal amount, rate of interest, and number of years.**

<i>Test Data:</i> 

amt = 10000, int = 3.5, years = 7

<i>Expected Output:</i> 

12722.79

In [51]:
# `round()` & Function
def future_value(amt, rate, years):
    return round(amt * (1 + rate * 0.01) ** years, 2)

print(future_value(10000, 3.5, 7))

12722.79


In [52]:
# `round()` & No function
amt = 10000
rate = 3.5
years = 7
future_value = amt * (1 + 0.01 * rate) ** years
print(round(future_value, 2))

12722.79


**40. Write a Python program to calculate the distance between the points (x1, y1) and (x2, y2).**

In [53]:
# List
from math import sqrt

def euclidean_distance(p1, p2):
    return sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2)

print(euclidean_distance([4, 0], [6, 6]))

6.324555320336759


In [54]:
# Tuple
from math import sqrt

def euclidean_distance(p1, p2):
    return sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2)

print(euclidean_distance((4, 0), (6, 6)))

6.324555320336759


**41. Write a Python program to check whether a file exists.**

In [55]:
# [1] `os.path.isfile`
import os.path

print(os.path.isfile('python_basic.ipynb'))
print(os.path.isfile('main.py'))

True
False


In [56]:
# [2] `os.path.exists`
import os.path

print(os.path.exists('python_basic.ipynb'))
print(os.path.exists('main.py'))

True
False


In [57]:
# [3] `open` & try-except block
my_file = open('python_basic.ipynb')
try:
    my_file.close()
    print("File found!")
except FileNotFoundError:
    print("File not found!")

File found!


**42. Write a Python program to determine if a Python shell is executing in 32bit or 64bit mode on OS.**

In [58]:
# `struct.calcsize()`
import struct

print(struct.calcsize("P") * 8)

64


In [59]:
# `platform.architecture()`
import platform

print(platform.architecture()[0])

64bit


**43. Write a Python program to get OS name, platform and release information.**

In [60]:
# `os.name` & `platform.system()` & `platform.release()`
import os, platform

print("Name of the operating system:", os.name)
print("Name of the OS system:", platform.system())
print("Version of the operating system:", platform.release())

Name of the operating system: posix
Name of the OS system: Linux
Version of the operating system: 5.4.0-148-generic


In [61]:
# `sys.platform` & `sysconfig.get_platform()` & `platform.machine()` & `platform.architecture()`
import sys, sysconfig, os, platform

print("os.name                     ", os.name)
print("sys.platform                ", sys.platform)
print("platform.system()           ", platform.system())
print("sysconfig.get_platform()    ", sysconfig.get_platform())
print("platform.machine()          ", platform.machine())
print("platform.architecture()     ", platform.architecture())

os.name                      posix
sys.platform                 linux
platform.system()            Linux
sysconfig.get_platform()     linux-x86_64
platform.machine()           x86_64
platform.architecture()      ('64bit', 'ELF')


**44. Write a Python program to locate Python site packages.**

In [62]:
# `site.getsitepackages()`
import site

print(site.getsitepackages())

['/home/yungshun/yungshun-py3/lib/python3.8/site-packages']


**45. Write a Python program that calls an external command.**

In [63]:
# [1] `subprocess.call()`
from subprocess import call

call(["ls", "-l"])

total 56
-rw-rw-r-- 1 yungshun yungshun 52993  五  17 12:14 python_basic.ipynb
-rw-rw-r-- 1 yungshun yungshun   503  五  17 10:24 python_basic.py


0

In [64]:
# [2] `os.system()`
import os

print(os.system("ls -l"))

total 56
-rw-rw-r-- 1 yungshun yungshun 52993  五  17 12:14 python_basic.ipynb
-rw-rw-r-- 1 yungshun yungshun   503  五  17 10:24 python_basic.py
0


**46. Write a Python program to retrieve the path and name of the file currently being executed.**

In [65]:
# [1] `os.path.realpath(__file__)`, `os.path.abspath(__file__)`, `sys.argv[0]`
import os, sys

# Not work in Jupyter Notebook
# print("Current File Name:", os.path.realpath(__file__))
# print("Current File Name:", os.path.abspath(__file__))
# print("Current File Name:", sys.argv[0])

# Output
# Current File Name: /home/yungshun/workspace/py/python-w3resource-exercises/main.py
# Current File Name: /home/yungshun/workspace/py/python-w3resource-exercises/main.py
# Current File Name: main.py

In [66]:
%%javascript

IPython.notebook.kernel.execute("nb_name = '" + IPython.notebook.notebook_name + "'")

<IPython.core.display.Javascript object>

In [67]:
# [2] Trick only for Jupyter Notebook & `os.getcwd()`
import os

nb_full_path = os.path.join(os.getcwd(), nb_name)
print(nb_full_path)

/home/yungshun/workspace/py/python-w3resource-exercises/python_basic.ipynb


**47. Write a Python program to find out the number of CPUs used.**

In [68]:
# [1] `multiprocessing.cpu_count()`
import multiprocessing

print(multiprocessing.cpu_count())

12


In [69]:
# [2] `os.cpu_count()`
import os

print(os.cpu_count())

12


In [70]:
# [3] `psutil`
import psutil

print(psutil.cpu_count())

12


**48. Write a Python program to parse a string to float or integer.**

In [71]:
# Try-except block & `ValueError` & `int()` & `float()`
def parse_string(s):
    try:
        return int(s)
    except ValueError:
        return float(s)

print(parse_string("3.14159"))

3.14159


**49. Write a Python program to list all files in a directory.**

In [72]:
# List initialization & `os.listdir()` & `os.path.isfile()` & `os.path.join()`
from os import listdir
from os.path import isfile, join

file_list = [f for f in listdir("/home/yungshun/workspace/py/python-w3resource-exercises") if isfile(join("/home/yungshun/workspace/py/python-w3resource-exercises", f))]
print(file_list)

['python_basic.ipynb', '.gitignore', 'python_basic.py']


**50. Write a Python program to print without a newline or space.**

In [73]:
# `print(objects, end="")`
for i in range(0, 10):
    print("*", end="")

**********

**51. Write a Python program to determine the profiling of Python programs.**

<i>Note:</i>

A profile is a set of statistics that describes how often and for how long various parts of the program executed. These statistics can be formatted into reports via the pstats module.

In [74]:
# `cProfile` module
import cProfile

def sum_numbers(x, y):
    print(x + y)
    
cProfile.run("sum_numbers(3, 5)")

8
         27 function calls in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 371060289.py:4(sum_numbers)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 iostream.py:202(schedule)
        2    0.000    0.000    0.000    0.000 iostream.py:429(_is_master_process)
        2    0.000    0.000    0.000    0.000 iostream.py:448(_schedule_flush)
        2    0.000    0.000    0.000    0.000 iostream.py:518(write)
        1    0.000    0.000    0.000    0.000 iostream.py:90(_event_pipe)
        1    0.000    0.000    0.000    0.000 socket.py:543(send)
        1    0.000    0.000    0.000    0.000 threading.py:1017(_wait_for_tstate_lock)
        1    0.000    0.000    0.000    0.000 threading.py:1071(is_alive)
        1    0.000    0.000    0.000    0.000 threading.py:513(is_set)
        1    0.000    0.000    0.000

**52. Write a Python program to print to STDERR.**

In [75]:
# `*args` & `**kwargs` & `print(objects, file=arg, sep=arg)` & `sys.stderr`
import sys
from __future__ import print_function

def eprint(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)
    
eprint("abc", "efg", "xyz", sep="--")

abc--efg--xyz


**53. Write a Python program to access environment variables.**

In [76]:
# `os.environ`
import os

print(os.environ)
print(os.environ["HOME"])
print(os.environ["PATH"])

environ({'CLUTTER_IM_MODULE': 'xim', 'CONDA_SHLVL': '1', 'LS_COLORS': 'rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.

**54. Write a Python program to get the current username.**

In [77]:
# [1] `getpass` module
import getpass

print(getpass.getuser())

yungshun


In [78]:
# [2] `pwd.getpwuid()` & `os.getuid()`
import os, pwd

def get_username():
    # `pwd.getpwuid()` gets username from uid
    return pwd.getpwuid(os.getuid())[0]

print(get_username())

yungshun


**55. Write a Python program to find local IP addresses using Python's stdlib.**

In [79]:
# `socket` module
import socket

print(
    [
        l for l in (
            [
                ip for ip in socket.gethostbyname_ex(socket.gethostname())[2] if not ip.startswith("127.")
            ][:1],
            [
                [
                    (s.connect(("8.8.8.8", 53)), s.getsockname()[0], s.close()) for s in [socket.socket(socket.AF_INET, socket.SOCK_DGRAM)]
                ][0][1]
            ]
        ) if l
    ][0][0]
)

192.168.169.106


**56. Write a Python program to get the height and width of the console window.**

In [80]:
# [1] `fcntl` & `termios` & `struct` modules
import fcntl, termios, struct

# Not work in Jupyter Notebook
# def terminal_size():
#     th, tw, hp, wp = struct.unpack("HHHH", 
#                                    fcntl.ioctl(0, termios.TIOCGWINSZ,
#                                    struct.pack("HHHH", 0, 0, 0, 0)))
#     return tw, th

# print("Number of columns and rows:", terminal_size())

# Output
# Number of columns and rows: (167, 42)

In [81]:
%%javascript 

var outerWidth = window.outerWidth
var outerHeight = window.outerHeight
IPython.notebook.kernel.execute("outer_width = " + outerWidth)
IPython.notebook.kernel.execute("outer_height = " + outerHeight)

<IPython.core.display.Javascript object>

In [82]:
# [2] `%%javascript` & `IPython.notebook.kernel.execute()` only for Jupyter Notebook
print(outer_width)
print(outer_height)

1853
1053


**57. Write a Python program to get the execution time of a Python method.**

In [83]:
# [1] Pass Python function as a function argument & `time.time()`
import time

def execution_time(func, n):
    start_time = time.time()
    func(n)
    end_time = time.time()
    return end_time - start_time

def sum_of_n_numbers(n):
    result = 0
    for i in range(1, n + 1):
        result = result + i
    print("Sum of 1 to", n, "is:", result)

n = 100
print("Required time to sum of 1 to", n, "is:", execution_time(sum_of_n_numbers, n))

Sum of 1 to 100 is: 5050
Required time to sum of 1 to 100 is: 2.956390380859375e-05


In [84]:
# [2] `time.time()`
import time

def sum_of_n_numbers(n):
    start_time = time.time()
    result = 0
    for i in range(1, n + 1):
        result = result + i
    end_time = time.time()
    return result, end_time - start_time

n = 100
print("Time to sum of 1 to", n, "and required time to calculate is:", sum_of_n_numbers(n))

Time to sum of 1 to 100 and required time to calculate is: (5050, 3.337860107421875e-06)


**58. Write a Python program to sum the first n positive integers.**

In [85]:
# [1] Function & `range()`
def sum_first_n(n):
    result = 0
    for i in range(1, n + 1):
        result = result + i
    return result

n = 100
print("Sum of first", n, "positive integers is", sum_first_n(n))

Sum of first 100 positive integers is 5050


In [86]:
# [2] `input()` & `range()`
n = int(input("Input a number: "))

sum_num = (n * (n + 1)) / 2

print("Sum of the first", n ,"positive integers:", sum_num)

Input a number: 100
Sum of the first 100 positive integers: 5050.0


In [87]:
# [3] `sum()` & `input()` & `range()` 
n = int(input("Input a number: "))

sum_num = sum(range(n + 1))

print("Sum of the first", n ,"positive integers:", sum_num)

Input a number: 100
Sum of the first 100 positive integers: 5050


**59. Write a Python program to convert height (in feet and inches) to centimeters.**

In [92]:
# [1] Function
def height_to_cm(h_ft, h_inch):
    return h_ft * 30.48 + h_inch * 2.54

print(height_to_cm(5, 9))

175.26


In [93]:
# [2] `input()`
print("Input your height:")
h_ft = int(input("Feet: "))
h_inch = int(input("Inch: "))

h_inch = h_inch + h_ft * 12
h_cm = round(h_inch * 2.54, 1)

print("Your height is : %d cm." % h_cm)

Input your height:
Feet: 5
Inch: 9
Your height is : 175 cm.


**60. Write a Python program to calculate the hypotenuse of a right angled triangle.**

In [95]:
# [1] Function
from math import sqrt

def hypotenuse(a, b):
    return sqrt(a**2 + b**2)

print(hypotenuse(3, 4))

5.0


In [96]:
# [2] `input()`
from math import sqrt

print("Input lengths of shorter triangle sides:")
a = float(input("a: "))
b = float(input("b: "))

c = sqrt(a**2 + b**2)

print("The length of the hypotenuse is:", c)

Input lengths of shorter triangle sides:
a: 3
b: 4
The length of the hypotenuse is: 5.0


**61. Write a Python program to convert the distance (in feet) to inches, yards, and miles.**

In [98]:
# `%.2f`
d_ft = int(input("Input distance in feet: "))

d_inches = d_ft * 12
d_yards = d_ft / 3.0
d_miles = d_ft / 5280.0

print("The distance in inches is %i inches." % d_inches)
print("The distance in yards is %.2f yards." % d_yards)
print("The distance in miles is %.2f miles." % d_miles)

Input distance in feet: 100
The distance in inches is 1200 inches.
The distance in yards is 33.33 yards.
The distance in miles is 0.02 miles.


**62. Write a Python program to convert all units of time into seconds.**

In [100]:
# `input()`
days = int(input("Input days: ")) * 3600 * 24
hours = int(input("Input hours: ")) * 3600
minutes = int(input("Input minutes: ")) * 60
seconds = int(input("Input seconds: "))

time = days + hours + minutes + seconds

print("The amounts of seconds is", time)

Input days: 4
Input hours: 5
Input minutes: 20
Input seconds: 10
The amounts of seconds is 364810


**63. Write a Python program to get an absolute file path.**

In [103]:
# [1] `os.path.abspath()`
import os

print("Absolute file path:", os.path.abspath("python_basic.ipynb"))

Absolute file path: /home/yungshun/workspace/py/python-w3resource-exercises/python_basic.ipynb


In [104]:
# [2] `pathlib.Path(param).resolve()`
from pathlib import Path

p = Path("python_basic.ipynb").resolve()

print(p)

/home/yungshun/workspace/py/python-w3resource-exercises/python_basic.ipynb


**64. Write a Python program that retrieves the date and time of file creation and modification.**

In [105]:
# `time.ctime()` & `os.path.getmtime()` & `os.path.getctime()`
import os, time

print("Last modified: %s" % time.ctime(os.path.getmtime("python_basic.ipynb")))
# `getctime()` only returns the creation time on Windows
# `getctime()` returns the time of the last change in Unix
print("Created: %s" % time.ctime(os.path.getctime("python_basic.ipynb")))

Last modified: Wed May 17 12:46:45 2023
Created: Wed May 17 12:46:45 2023


**65. Write a Python program that converts seconds into days, hours, minutes, and seconds.**

In [114]:
# [1] `//` & `%` & Function 
def seconds_to_dhms(time):
    days = time // (3600 * 24)
    time = time % (3600 * 24)
    hours = time // 3600
    time = time % 3600
    minutes = time // 60
    seconds = time % 60
    return days, hours, minutes, seconds

print("d:h:m:s -> %d:%d:%d:%d" % seconds_to_dhms(1234567))

d:h:m:s -> 14:6:56:7


In [115]:
# [2] `//` & `%` & `input()`
time = int(input("Input time in seconds: "))

days = time // (3600 * 24)
time = time % (3600 * 24)
hours = time // 3600
time = time % 3600
minutes = time // 60
seconds = time % 60

print("d:h:m:s -> %d:%d:%d:%d" % (days, hours, minutes, seconds))

Input time in seconds: 1234567
d:h:m:s -> 14:6:56:7


**66. Write a Python program to calculate the body mass index.**

In [117]:
# `input()`
height = float(input("Input your height in meter: "))
weight = float(input("Input your weight in kilogram: "))

print("Your body mass index is:", round(weight / (height**2), 2))

Input your height in meter: 1.76
Input your weight in kilogram: 67.9
Your body mass index is: 21.92


**67. Write a Python program to convert pressure in kilopascals to pounds per square inch, a millimeter of mercury (mmHg) and atmosphere pressure.**

In [119]:
# `input()`
kpa = float(input("Input pressure in kilopascals: "))

psi = kpa / 6.89475729
mmhg = kpa * 760 / 101.325
atm = kpa / 101.325

print("The pressure in pounds per square inch: %.2f psi"  % (psi))
print("The pressure in millimeter of mercury: %.2f mmHg" % (mmhg))
print("Atmosphere pressure: %.2f atm" % (atm))

Input pressure in kilopascals: 12.35
The pressure in pounds per square inch: 1.79 psi
The pressure in millimeter of mercury: 92.63 mmHg
Atmosphere pressure: 0.12 atm


**68. Write a Python program to calculate sum of digits of a number.**

In [122]:
# Loop iteration & `%` & `//`
def sum_of_digits(n):
    result = 0
    while n != 0:
        result = result + n % 10
        n = n // 10
    return result

print(sum_of_digits(1234567))
print(sum_of_digits(8))
print(sum_of_digits(317))

28
8
11


**69. Write a Python program to sort three integers without using conditional statements and loops.**

In [125]:
# Trick using `min()` & `max()` & calculation
x = int(input("Input first number: "))
y = int(input("Input second number: "))
z = int(input("Input third number: "))

a1 = min(x, y, z)
a3 = max(x, y, z)
a2 = (x + y + z) - a1 - a3

print("Numbers in sorted order:", a1, a2, a3)

Input first number: 5
Input second number: 8
Input third number: 3
Numbers in sorted order: 3 5 8


**70. Write a Python program to sort files by date.**

In [126]:
# [1] `glob` module
import os, glob

files = glob.glob("*")
files.sort(key=os.path.getmtime)

print("\n".join(files))

python_basic.py
python_basic.ipynb


In [128]:
# [2] `os.chdir()` & `os.listdir()` & `os.path.isfile()` & `filter()` & `map()`
import os

os.chdir("/home/yungshun/workspace/py/python-w3resource-exercises")
result = sorted(filter(os.path.isfile, os.listdir(".")), key=os.path.getmtime)

print("\n".join(map(str, result)))

.gitignore
python_basic.py
python_basic.ipynb


**71. Write a Python program to get a directory listing, sorted by creation date.**

In [3]:
# [1] List declaration & `os.path.getctime()` & `os.listdir()`
import os, time

paths = ["%s %s" % (time.ctime(t), f) for t, f in sorted([(os.path.getctime(x), x) for x in os.listdir(".")])]

print("Directory listing, sorted by creation date:")
for x in range(len(paths)):
    print(paths[x],)

Directory listing, sorted by creation date:
Tue May 16 04:22:18 2023 .ipynb_checkpoints
Tue May 16 08:44:47 2023 .gitignore
Wed May 17 13:42:34 2023 python_basic.py
Wed May 17 13:43:27 2023 .git
Tue May 23 22:32:41 2023 python_basic.ipynb


In [4]:
# [2] List declaration & `os.path.getctime()` & `glob.glob()`
import os, glob

files = ["%s %s" % (time.ctime(t), f) for t, f in sorted([(os.path.getctime(x), x) for x in glob.glob("*")])]

print("Directory listing, sorted by creation date:")
for x in range(len(files)):
    print(files[x],)

Directory listing, sorted by creation date:
Wed May 17 13:42:34 2023 python_basic.py
Tue May 23 22:34:41 2023 python_basic.ipynb


In [10]:
# [3] `stat` module & `os.stat()` & `sys.argv` & `time.ctime()`
import os, sys, time
from stat import S_ISREG, ST_CTIME, ST_MODE

# Relative or absolute path to the directory
dir_path = sys.argv[1] if len(sys.argv) == 2 else r'.'
# All entries in the directory w/ stats
data = (os.path.join(dir_path, fn) for fn in os.listdir(dir_path))
data = ((os.stat(path), path) for path in data)
# The `stat` module defines constants and functions for interpreting the results of `os.stat()`
# Regular files, insert creation date
data = ((stat[ST_CTIME], path) for stat, path in data if S_ISREG(stat[ST_MODE]))

for cdate, path in sorted(data):
    # print(cdate, os.path.basename(path))
    # 1684197887 .gitignore
    # 1684302154 python_basic.py
    # 1684853441 python_basic.ipynb

    # `ctime()` converts a time expressed in seconds
    print(time.ctime(cdate), os.path.basename(path))

Tue May 16 08:44:47 2023 .gitignore
Wed May 17 13:42:34 2023 python_basic.py
Tue May 23 22:50:41 2023 python_basic.ipynb


**72. Write a Python program to get the details of the math module.**

In [11]:
# [1] `dir()`
import math

# Sets everything to a list from the `math` module
math_ls = dir(math)

print(math_ls)

['__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'comb', 'copysign', 'cos', 'cosh', 'degrees', 'dist', 'e', 'erf', 'erfc', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'isqrt', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'perm', 'pi', 'pow', 'prod', 'radians', 'remainder', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc']


In [12]:
# [2] `help()`
import math

print("Details of math module:\n")

help(math)

Details of math module:

Help on module math:

NAME
    math

MODULE REFERENCE
    https://docs.python.org/3.8/library/math
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    This module provides access to the mathematical functions
    defined by the C standard.

FUNCTIONS
    acos(x, /)
        Return the arc cosine (measured in radians) of x.
    
    acosh(x, /)
        Return the inverse hyperbolic cosine of x.
    
    asin(x, /)
        Return the arc sine (measured in radians) of x.
    
    asinh(x, /)
        Return the inverse hyperbolic sine of x.
    
    atan(x, /)
        Return the arc tangent (measured in radians) of x.
    
    atan2(y, x, /)
        Return the arc tangent (measured

**73. Write a Python program to calculate the midpoints of a line.**

In [15]:
# `input()`
print("Calculate the midpoint of a line:")
x1 = float(input("The value of x (the first endpoint)"))
y1 = float(input("The value of y (the first endpoint)"))
x2 = float(input("The value of x (the second endpoint)"))
y2 = float(input("The value of y (the second endpoint)"))

x_m_point = (x1 + x2) / 2
y_m_point = (y1 + y2) / 2

print("The midpoint of line is:")
print("The midpoint's x value is:", x_m_point)
print("The midpoint's y value is:", y_m_point)

Calculate the midpoint of a line:
The value of x (the first endpoint)2
The value of y (the first endpoint)2
The value of x (the second endpoint)4
The value of y (the second endpoint)4
The midpoint of line is:
The midpoint's x value is: 3.0
The midpoint's y value is: 3.0


**74. Write a Python program to hash a word.**

In [17]:
# [1] `ord()`
soundex = [0,1,2,3,0,1,2,0,0,2,2,4,5,5,0,1,2,6,2,3,0,1,0,2,0,2]

word = input("Input the word be hashed: ")

word = word.upper()
coded = word[0]
for a in word[1:len(word)]:
    i = 65 - ord(a)
    coded = coded + str(soundex[i])

print("The coded word is:", coded)

Input the word be hashed: yungshun
The coded word is: Y2500325


In [20]:
# [2] `hash()`
int_num = 112
flt_num = 23.99
text_val = "Python Exercises"

print("Original integer value:", int_num)
print ("Hash value of the said integer value:", str(hash(int_num)))
print("Original float value:", flt_num)
print ("Hash value of the said float value:", str(hash(flt_num)))
print("Original text:", text_val)
print ("Hash value of the said text:", str(hash(text_val)))

Original integer value: 112
Hash value of the said integer value: 112
Original float value: 23.99
Hash value of the said float value: 2282784579121553431
Original text: Python Exercises
Hash value of the said text: -5274652165466088172


**75. Write a Python program to get the copyright information and write Copyright information in Python code.**

In [25]:
# [1] `sys.copyright`
import sys

print("Python Copyright Information:\n")
print(sys.copyright)

Python Copyright Information:

Copyright (c) 2001-2021 Python Software Foundation.
All Rights Reserved.

Copyright (c) 2000 BeOpen.com.
All Rights Reserved.

Copyright (c) 1995-2001 Corporation for National Research Initiatives.
All Rights Reserved.

Copyright (c) 1991-1995 Stichting Mathematisch Centrum, Amsterdam.
All Rights Reserved.


In [26]:
# [2] `copyright`
print("Python Copyright Information:\n")
print(copyright)

Python Copyright Information:

Copyright (c) 2001-2021 Python Software Foundation.
All Rights Reserved.

Copyright (c) 2000 BeOpen.com.
All Rights Reserved.

Copyright (c) 1995-2001 Corporation for National Research Initiatives.
All Rights Reserved.

Copyright (c) 1991-1995 Stichting Mathematisch Centrum, Amsterdam.
All Rights Reserved.


**76. Write a Python program to get the command-line arguments (name of the script, the number of arguments, arguments) passed to a script.**

In [27]:
# `sys.argv`
import sys

print("This is the name/path of the script:", sys.argv[0])
print("Number of arguments:", len(sys.argv))
print("Argument list:", str(sys.argv))

# ~$ python python_basic.py arg1 arg2 arg3 
# This is the name/path of the script: python_basic.py
# Number of arguments: 4
# Argument list: ['python_basic.py', 'arg1', 'arg2', 'arg3']

This is the name/path of the script: /home/yungshun/yungshun-py3/lib/python3.8/site-packages/ipykernel_launcher.py
Number of arguments: 3
Argument list: ['/home/yungshun/yungshun-py3/lib/python3.8/site-packages/ipykernel_launcher.py', '-f', '/home/yungshun/.local/share/jupyter/runtime/kernel-ba795fe6-3b28-4fe7-8754-b2b28549483b.json']


**77. Write a Python program to test whether the system is a big-endian platform or a little-endian platform.**

In [29]:
# `sys.byteorder`
import sys

if sys.byteorder == "little":
    # intel, alpha
    print("Little-endian platform")
else:
    # motorola, sparc
    print("Big-endian platform")

Little-endian platform


**78. Write a Python program to find the available built-in modules.**

In [30]:
# [1] `sys.builtin_module_names` & `textwrap.fill()`
import sys, textwrap

module_name = ", ".join(sorted(sys.builtin_module_names))

# Wraps the single paragraph in text & returns a single string containing the wrapped paragraph
print(textwrap.fill(module_name, width=70))

_abc, _ast, _codecs, _collections, _functools, _imp, _io, _locale,
_operator, _signal, _sre, _stat, _string, _symtable, _thread,
_tracemalloc, _warnings, _weakref, atexit, builtins, errno,
faulthandler, gc, itertools, marshal, posix, pwd, sys, time, xxsubtype


In [31]:
# [2] `help()`
help("modules")


Please wait a moment while I gather a list of all available modules...



2023-05-24 13:56:29.215435: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-05-24 13:56:29.737304: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2023-05-24 13:56:29.737352: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2023-05-24 13:56:29.814715: E tensorflow/stream_executor/cuda/cuda_blas.cc:2981] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-05-24 13:56:31.005644: W tensorflow/stream_executor/platform/de

IPython             atexit              itertools           reprlib
PIL                 attr                jedi                requests
__future__          attrs               jinja2              requests_oauthlib
_abc                audioop             joblib              resource
_argon2_cffi_bindings backcall            json                rlcompleter
_ast                base64              jsonschema          rsa
_asyncio            bdb                 jupyter             runpy
_bisect             binascii            jupyter_client      sched
_blake2             binhex              jupyter_core        scipy
_bootlocale         bisect              jupyterlab_plotly   scs
_bz2                bleach              jupyterlab_pygments seaborn
_cffi_backend       bs4                 keras               secrets
_codecs             builtins            keras_preprocessing select
_codecs_cn          bz2                 keyword             selectors
_codecs_hk          cProfile            kiw

/home/yungshun/anaconda3/lib/python3.8/pkgutil.py:107: VisibleDeprecationWarning:

zmq.eventloop.minitornado is deprecated in pyzmq 14.0 and will be removed.
    Install tornado itself to use zmq with the tornado IOLoop.
    



**79. Write a Python program to get the size of an object in bytes.**

In [37]:
# `sys.getsizeof()`
import sys

str1 = "one"
str2 = "two"
str3 = "three"
x = 0
y = 112
z = 122.56
L = [1, 2, 3, "Red", "Black"]
T = ("Red", [8, 4, 6], (1, 2, 3))
S = {"apple", "orange", "apple", "pear"}
D = {"Name": "David", "Age": 6, "Class": "First"}

print("Size of", str1, "=", str(sys.getsizeof(str1)), "bytes")
print("Size of", str2, "=", str(sys.getsizeof(str2)), "bytes")
print("Size of", str3, "=", str(sys.getsizeof(str3)), "bytes")
print("Size of", x, "=", str(sys.getsizeof(x)), "bytes")
print("Size of", y, "=", str(sys.getsizeof(y)), "bytes") 
print("Size of", L, "=", str(sys.getsizeof(L)), "bytes")
print("Size of", T, "=", str(sys.getsizeof(T)), "bytes")
print("Size of", S, "=", str(sys.getsizeof(S)), "bytes")
print("Size of", D, "=", str(sys.getsizeof(D)), "bytes")

Size of one = 52 bytes
Size of two = 52 bytes
Size of three = 54 bytes
Size of 0 = 24 bytes
Size of 112 = 28 bytes
Size of [1, 2, 3, 'Red', 'Black'] = 96 bytes
Size of ('Red', [8, 4, 6], (1, 2, 3)) = 64 bytes
Size of {'apple', 'orange', 'pear'} = 216 bytes
Size of {'Name': 'David', 'Age': 6, 'Class': 'First'} = 232 bytes


**80. Write a Python program to get the current value of the recursion limit.**

In [38]:
# [1] `sys.getrecursionlimit()`
import sys

print("Current value of the recursion limit:")
print(sys.getrecursionlimit())

Current value of the recursion limit:
3000


In [39]:
# [2] `sys.getrecursionlimit()` & `sys.setrecursionlimit()`
import sys

print("Call sys.getrecursionlimit() to get the current recursion limit:")
recursion_limit = sys.getrecursionlimit()
print(recursion_limit)

print("Call sys.setrecursionlimit(n) to change the recursion limit to n operations:")
sys.setrecursionlimit(1000)
new_recursion_limit = sys.getrecursionlimit()
print(new_recursion_limit)

Call sys.getrecursionlimit() to get the current recursion limit:
3000
Call sys.setrecursionlimit(n) to change the recursion limit to n operations:
1000


**81. Write a Python program to concatenate N strings.**

In [40]:
# [1] `join()`
list_of_colors = ["Red", "Green", "Blue"]
colors = "-".join(list_of_colors)

print("All colors:", colors)

All colors: Red-Green-Blue


In [44]:
# [2] `+`
print("Concatenete with the `+` operator:")

list_of_colors = ["Red", "Green", "Blue"]
colors = list_of_colors[0] + "-" + list_of_colors[1] + "-" + list_of_colors[2]

print("All colors:", colors)

Concatenete with the `+` operator:
All colors: Red-Green-Blue


**82. Write a Python program to calculate the sum of all items of a container (tuple, list, set, dictionary).**

In [74]:
# [1] List container & `sum()`
nums = [10, 20, 30]

print("Original container:")
print(nums)
print(type(nums))
print("Sum of all items of the said container:", sum(nums))

Original container:
[10, 20, 30]
<class 'list'>
Sum of all items of the said container: 60


In [76]:
# [2] Tuple container & `sum()`
nums = (10, 20, 30)

print("Original container:")
print(nums)
print(type(nums))
print("Sum of all items of the said container:", sum(nums))

Original container:
(10, 20, 30)
<class 'tuple'>
Sum of all items of the said container: 60


In [78]:
# [3] Tuple container & `sum()`
nums = {10, 20, 30}

print("Original container:")
print(nums)
print(type(nums))
print("Sum of all items of the said container:", sum(nums))

Original container:
{10, 20, 30}
<class 'set'>
Sum of all items of the said container: 60


In [79]:
# [4] Dictionary container & `sum()`
def dict_sum(nums):
    num_sum = 0
    for i in nums:
        num_sum = num_sum + nums[i]
    return num_sum

nums = {"a": 10, "b": 20, "c": 30}

print("Original container:")
print(nums)
print(type(nums))
print("Sum of all items of the said container:", dict_sum(nums))

Original container:
{'a': 10, 'b': 20, 'c': 30}
<class 'dict'>
Sum of all items of the said container: 60


In [81]:
# [5] Dictionary container & `values()` & `sum()`
nums = {"a": 10, "b": 20, "c": 30}

print("Original container:")
print(nums)
print(type(nums))
print("Sum of all items of the said container:", sum(nums.values()))

Original container:
{'a': 10, 'b': 20, 'c': 30}
<class 'dict'>
Sum of all items of the said container: 60


In [82]:
# [5] List, tuple, set containers & `reduce()`
import functools

# Also works for tuple & set containers 
nums = [10, 20, 30]

print("Original container:")
print(nums)
print(type(nums))
print("Sum of all items of the said container:", functools.reduce(lambda a, b: a + b, nums))

Original container:
[10, 20, 30]
<class 'list'>
Sum of all items of the said container: 60


In [83]:
# [6] Dictionary containers & `values()` & `reduce()`
import functools

# Also works for tuple & set containers 
nums = {"a": 10, "b": 20, "c": 30}

print("Original container:")
print(nums)
print(type(nums))
print("Sum of all items of the said container:", functools.reduce(lambda a, b: a + b, nums.values()))

Original container:
{'a': 10, 'b': 20, 'c': 30}
<class 'dict'>
Sum of all items of the said container: 60


**83. Write a Python program to test whether all numbers in a list are greater than a certain number.**

In [67]:
# `all()`
# `all()` returns `True` if all items in an iterable are true; otherwise it returns `False`
def gt(nums, n):
    return all(x > n for x in nums)

nums = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
print("Original list numbers:")
print(nums)

n = 5
print("Check whether all numbers of the said list greater than:", n)
print(gt(nums, n))

n = 13
print("Check whether all numbers of the said list greater than:", n)
print(gt(nums, n))

Original list numbers:
[10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
Check whether all numbers of the said list greater than: 5
True
Check whether all numbers of the said list greater than: 13
False


**84. Write a Python program to count the number of occurrences of a specific character in a string.**

In [87]:
# [1] Loop
def word_count(s, ch):
    count = 0
    for c in s:
        if c == ch: 
            count = count + 1
    return count

s = "The quick brown fox jumps over the lazy dog."
ch = "o"

print("Original string:")
print(s)
print("Number of occurrence of", "\"" + ch + "\"", "in the said string:")
print(word_count(s, ch))

Original string:
The quick brown fox jumps over the lazy dog.
Number of occurrence of "o" in the said string:
4


In [85]:
# [2] `count()`
s = "The quick brown fox jumps over the lazy dog."
ch = "o"

print("Original string:")
print(s)
print("Number of occurrence of", "\"" + ch + "\"", "in the said string:")
print(s.count(ch))

Original string:
The quick brown fox jumps over the lazy dog.
Number of occurrence of "o" in the said string:
4


In [88]:
# [3] `collections.Counter()`
from collections import Counter

s = "The quick brown fox jumps over the lazy dog."
ch = "o"

print("Original string:")
print(s)
print("Number of occurrence of", "\"" + ch + "\"", "in the said string:")
print(str(Counter(s)[ch]))

Original string:
The quick brown fox jumps over the lazy dog.
Number of occurrence of "o" in the said string:
4


In [90]:
# [4] `lambda` & `map()` & `sum()`
s = "The quick brown fox jumps over the lazy dog."
ch = "o"

print("Original string:")
print(s)
print("Number of occurrence of", "\"" + ch + "\"", "in the said string:")
print(sum(map(lambda x: 1 if ch in x else 0, s)))

Original string:
The quick brown fox jumps over the lazy dog.
Number of occurrence of "o" in the said string:
4


In [91]:
# [5] `re.findall()`
import re

s = "The quick brown fox jumps over the lazy dog."
ch = "o"

print("Original string:")
print(s)
print("Number of occurrence of", "\"" + ch + "\"", "in the said string:")
print(len(re.findall(ch, s)))

Original string:
The quick brown fox jumps over the lazy dog.
Number of occurrence of "o" in the said string:
4


**85. Write a Python program to check whether a file path is a file or a directory.**

In [93]:
# `os.isdir()` & `os.isfile()`
import os

path = "python_basic.ipynb"

if os.path.isdir(path):
    print("It is a directory")
elif os.path.isfile(path):
    print("It is a normal file")
else:
    print("It is a special file (socket, FIFO, device file)")

It is a normal file


**86. Write a Python program to get the ASCII value of a character.**

In [94]:
# `ord()`
print(ord("a"))
print(ord("A"))
print(ord("1"))
print(ord("@"))

97
65
49
64


**87. Write a Python program to get the size of a file.**

In [107]:
# [1] `os.path.getsize()`
import os

file_size = os.path.getsize("python_basic.ipynb")

print("The size of the file is:", file_size, "Bytes")

The size of the file is: 139721 Bytes


In [108]:
# [2] `os.stat().st_size`
import os

file_size = os.stat("python_basic.ipynb").st_size

print("The size of the file is:", file_size, "Bytes")

The size of the file is: 139721 Bytes


In [109]:
# [3] `pathlib.Path()` & `Path.stat().st_size`
import pathlib

file = pathlib.Path("python_basic.ipynb")
file_size = file.stat().st_size

print("The size of the file is:", file_size, "Bytes")

The size of the file is: 139721 Bytes


In [110]:
# [4] `open()` & `seek()` & `tell()`
import os

with open("python_basic.ipynb", "rb") as file:
    file.seek(0, os.SEEK_END)
    print("The size of the file is:", file.tell(), "Bytes")

The size of the file is: 139721 Bytes


**88. Given variables x=30 and y=20, write a Python program to print "30+20=50".**

In [111]:
# [1] `%d`
x = 30
y = 20

print("%d+%d=%d" % (x, y, x + y))

30+20=50


In [112]:
# [2] `{index}`
x = 30
y = 20

print("{0}+{1}={2}".format(x, y, x + y))

30+20=50


In [113]:
# [3] `{}`
x = 30
y = 20

print("{}+{}={}".format(x, y, x + y))

30+20=50


**89. Write a Python program to perform an action if a condition is true.**

Given a variable name, if the value is 1, display the string "First day of a Month!" and do nothing if the value is not equal.

In [114]:
# [1] if-else statement
n = 1

if n == 1:
    print("First day of a Month!")

First day of a Month!


In [115]:
# [2] `input` & if-else statement
n = float(input("Input a number: "))

if n == 1:
    print("First day of a Month!")

Input a number: 1
First day of a Month!


**90. Write a Python program to create a copy of its own source code.**

In [118]:
# `open()` & `write()` & `read()`
def file_copy(src, dest):
    with open(src) as f, open(dest, "w") as d:
        d.write(f.read())
        
file_copy("python_basic.py", "python_basic_copy.py")
        
with open("python_basic_copy.py", "r") as filehandle:
    for line in filehandle:
        print(line, end="")

# Problem 46
import os, sys

print("Current File Name:", os.path.realpath(__file__))
print("Current File Name:", os.path.abspath(__file__))
print("Current File Name:", sys.argv[0])

# Problem 56
import fcntl, termios, struct

def terminal_size():
    th, tw, hp, wp = struct.unpack("HHHH", 
                                   fcntl.ioctl(0, termios.TIOCGWINSZ,
                                   struct.pack("HHHH", 0, 0, 0, 0)))
    return tw, th

print("Number of columns and rows:", terminal_size())

# Problem 76
import sys

print("This is the name/path of the script:", sys.argv[0])
print("Number of arguments:", len(sys.argv))
print("Argument list:", str(sys.argv))


**91. Write a Python program to swap two variables.**

In [121]:
# [1] Without using temporary variable
a = 30
b = 20

print("Before swapping: a = %d and b = %d" % (a, b))

a, b = b, a

print("After swapping: a = %d and b = %d" % (a, b))

Before swapping: a = 30 and b = 20
After swapping: a = 20 and b = 30


In [123]:
# [2] Third temporary variable
a = 30
b = 20

print("Before swapping: a = %d and b = %d" % (a, b))

tmp = a
a = b
b = tmp

print("After swapping: a = %d and b = %d" % (a, b))

Before swapping: a = 30 and b = 20
After swapping: a = 20 and b = 30


**92. Write a Python program to define a string containing special characters in various forms.**

In [132]:
# Various methods for escaping `"` & `'` in string
# Use `\"` inside double quotes
print("\#{'}${\"}@/")
# Use `"'"'"` inside double quotes
print("\#{'}${"'"'"}@/")
# Wrapped by `r""" """`
print(r"""\#{'}${"}@/""")
# Use `\'` inside single quotes
print('\#{\'}${"}@/')
# Use `'"'"'` inside single quotes
print('\#{'"'"'}${"}@/')
# Wrapped by `r''' '''`
print(r'''\#{'}${"}@/''')

\#{'}${"}@/
\#{'}${"}@/
\#{'}${"}@/
\#{'}${"}@/
\#{'}${"}@/
\#{'}${"}@/


**93. Write a Python program to get the Identity, Type, and Value of an object.**

In [133]:
# `type()` & `id()`
x = 36

print("Identity:", x)
print("Type:", type(x))
print("Value:", id(x))

Identity: 36
Type: <class 'int'>
Value: 94209972281760


In [134]:
# [1] `b"str"`
x = b"Abc"

print("Convert bytes of the said string to a list of integer:", list(x))

Convert bytes of the said string to a list of integer: [65, 98, 99]


In [137]:
# [2] `ord()`
s = "Abc"
print("Original string:", s)

nums = []
for ch in s:
    nums.append(ord(ch))
    
print("Convert bytes of the said string to a list of integers:", nums)

Original string: Abc
Convert bytes of the said string to a list of integers: [65, 98, 99]


**95. Write a Python program to check whether a string is numeric.**

In [138]:
# [1] `isdigit()`
text = input("Input a word or number: ")

if text.isdigit():
    print("The input value is a number")
else:
    print("The input value is a string")

Input a word or number: yungshun
The input value is a string


In [139]:
# [2] try-except block
text = "yungshun"

try: 
    x = float(text)
except (ValueError, TypeError):
    print("Not numeric")

Not numeric


**96. Write a Python program to print the current call stack.**

In [140]:
# `traceback.print_stack()`
import traceback

def nested_func():
    return func()

def func():
    traceback.print_stack()
    
nested_func()

  File "/home/yungshun/anaconda3/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/yungshun/anaconda3/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/yungshun/yungshun-py3/lib/python3.8/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "/home/yungshun/yungshun-py3/lib/python3.8/site-packages/traitlets/config/application.py", line 978, in launch_instance
    app.start()
  File "/home/yungshun/yungshun-py3/lib/python3.8/site-packages/ipykernel/kernelapp.py", line 712, in start
    self.io_loop.start()
  File "/home/yungshun/yungshun-py3/lib/python3.8/site-packages/tornado/platform/asyncio.py", line 215, in start
    self.asyncio_loop.run_forever()
  File "/home/yungshun/anaconda3/lib/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/home/yungshun/anaconda3/lib/python3.8/asyncio/base_events.py", 

**97. Write a Python program to list the special variables used in the language.**

In [152]:
# `__builtins__.__dict__` & `globals()` & set operations
# s_var_names = sorted((set(globals().keys()) | set(__builtins__.__dict__.keys())) - set("_ names i".split()))
s_var_names = sorted(set(__builtins__.__dict__.keys()))

print("\n".join(" ".join(s_var_names[i:i+8]) for i in range(0, len(s_var_names), 8)))

ArithmeticError AssertionError AttributeError BaseException BlockingIOError BrokenPipeError BufferError BytesWarning
ChildProcessError ConnectionAbortedError ConnectionError ConnectionRefusedError ConnectionResetError DeprecationWarning EOFError Ellipsis
EnvironmentError Exception False FileExistsError FileNotFoundError FloatingPointError FutureWarning GeneratorExit
IOError ImportError ImportWarning IndentationError IndexError InterruptedError IsADirectoryError KeyError
KeyboardInterrupt LookupError MemoryError ModuleNotFoundError NameError None NotADirectoryError NotImplemented
NotImplementedError OSError OverflowError PendingDeprecationWarning PermissionError ProcessLookupError RecursionError ReferenceError
ResourceWarning RuntimeError RuntimeWarning StopAsyncIteration StopIteration SyntaxError SyntaxWarning SystemError
SystemExit TabError TimeoutError True TypeError UnboundLocalError UnicodeDecodeError UnicodeEncodeError
UnicodeError UnicodeTranslateError UnicodeWarning UserWarning 

**98. Write a Python program to get system time.**

<i>Note:</i>

The system time is important for debugging, network information, random number seeds, or something as simple as program performance.

In [153]:
# [1] `time.ctime()`
import time

print(time.ctime())

Wed May 24 16:47:05 2023


In [154]:
# [2] `datetime.datetime.now()`
import datetime

print(datetime.datetime.now())

2023-05-24 16:48:05.296189


**99. Write a Python program to clear the screen or terminal.**

In [156]:
# `os.system()`
import os, time

# Windows
# os.system("cls")
os.system("ls")
time.sleep(2)

# Ubuntu
os.system("clear")
# Works in terminal

python_basic_copy.py
python_basic.ipynb
python_basic.py


0

**100. Write a Python program to get the name of the host on which the routine is running.**

In [159]:
# [1] `socket.gethostname()`
import socket

host_name = socket.gethostname()

print("Host name:", host_name)

Host name: yungshun-GF63-Thin-10SCSR


In [160]:
# [2] `platform.uname()[1]`
import platform

host_name = platform.uname()[1]

print("Host name:", host_name)

Host name: yungshun-GF63-Thin-10SCSR


In [161]:
# [3] `os.uname().nodename`
import os

host_name = os.uname().nodename

print("Host name:", host_name)

Host name: yungshun-GF63-Thin-10SCSR


**101. Write a Python program to access and print a URL's content to the console.**

In [1]:
# `http.client.HTTPConnection()`
from http.client import HTTPConnection

conn = HTTPConnection("example.com")
conn.request("GET", "/")
result = conn.getresponse()
# Retrieve the entire contents
contents = result.read()

print(contents)

b'<!doctype html>\n<html>\n<head>\n    <title>Example Domain</title>\n\n    <meta charset="utf-8" />\n    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />\n    <meta name="viewport" content="width=device-width, initial-scale=1" />\n    <style type="text/css">\n    body {\n        background-color: #f0f0f2;\n        margin: 0;\n        padding: 0;\n        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;\n        \n    }\n    div {\n        width: 600px;\n        margin: 5em auto;\n        padding: 2em;\n        background-color: #fdfdff;\n        border-radius: 0.5em;\n        box-shadow: 2px 3px 7px 2px rgba(0,0,0,0.02);\n    }\n    a:link, a:visited {\n        color: #38488f;\n        text-decoration: none;\n    }\n    @media (max-width: 700px) {\n        div {\n            margin: 0 auto;\n            width: auto;\n        }\n    }\n    </style>    \n</head>\n\n<body>\n<div>\n    

**102. Write a Python program to get system command output.**

In [3]:
# `subprocess.check_output()`
import subprocess

# File & directory listing
returned_text = subprocess.check_output("dir", shell=True, universal_newlines=True)

print("dir command to list file and directory")
print(returned_text)

dir command to list file and directory
numpy_basic.ipynb     python_basic.py		    tmp_array.npy
python_basic_copy.py  python_data_types_list.ipynb  tmp_array.npz
python_basic.ipynb    python_lambda.ipynb



**103. Write a Python program to extract the filename from a given path.**

In [4]:
# `os.path.basename()`
import os 

print(os.path.basename("/home/yungshun/workspace/py/python-w3resource-exercises/python_basic.ipynb"))

python_basic.ipynb


**104. Write a Python program to get the effective group id, effective user id, real group id, and a list of supplemental group ids associated with the current process.**


<i>Note:</i> 

Availability: Unix.

In [5]:
# `os.getegid()` & `os.geteuid()` & `os.getgid()` & `os.getgroups()`
import os

print("Effective group id:", os.getegid())
print("Effective user id:", os.geteuid())
print("Real group id:", os.getgid())
print("List of supplemental group ids:", os.getgroups())

Effective group id: 1000
Effective user id: 1000
Real group id: 1000
List of supplemental group ids: [4, 24, 27, 30, 46, 116, 126, 127, 999, 1000]


**105. Write a Python program to get the users environment.**

In [6]:
# [1] `os.environ`
import os

print(os.environ)

environ({'CLUTTER_IM_MODULE': 'xim', 'CONDA_SHLVL': '1', 'LS_COLORS': 'rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.

In [7]:
# [2] `os.environ` & `pprint.pprint()`
import os, pprint

# User's environment variables
u_env_var = os.environ

print("User's environment variables:")
pprint.pprint(dict(u_env_var), width = 1)

User's environment variables:
{'ANDROID_HOME': '/home/yungshun/Android/Sdk',
 'ANDROID_SDK_ROOT': '/home/yungshun/Android/Sdk',
 'CLICOLOR': '1',
 'CLUTTER_IM_MODULE': 'xim',
 'COLORTERM': 'truecolor',
 'CONDA_DEFAULT_ENV': 'base',
 'CONDA_EXE': '/home/yungshun/anaconda3/bin/conda',
 'CONDA_PREFIX': '/home/yungshun/anaconda3',
 'CONDA_PROMPT_MODIFIER': '(base) ',
 'CONDA_PYTHON_EXE': '/home/yungshun/anaconda3/bin/python',
 'CONDA_SHLVL': '1',
 'DBUS_SESSION_BUS_ADDRESS': 'unix:path=/run/user/1000/bus',
 'DESKTOP_AUTOSTART_ID': '10ed186b9c6d65b81a168372379988255400000029040007',
 'DESKTOP_SESSION': 'ubuntu',
 'DISPLAY': ':0',
 'GDMSESSION': 'ubuntu',
 'GIT_PAGER': 'cat',
 'GNOME_DESKTOP_SESSION_ID': 'this-is-deprecated',
 'GNOME_SHELL_SESSION_MODE': 'ubuntu',
 'GNOME_TERMINAL_SCREEN': '/org/gnome/Terminal/screen/ff94611f_cb46_4ff1_b194_9efdfab22086',
 'GNOME_TERMINAL_SERVICE': ':1.152',
 'GPG_AGENT_INFO': '/run/user/1000/gnupg/S.gpg-agent:0:1',
 'GTK_IM_MODULE': 'ibus',
 'GTK_MODULES': 

**106. Write a Python program to divide a path by the extension separator.**

In [12]:
# `os.path.splitext()`
import os.path

for path in ["python_basic.ipynb", "filename", "/home/yungshun/workspace/py/python_w3resource_exercises/python_basic.ipynb", "/", ""]:
    print('"%s":' % path, os.path.splitext(path))

"python_basic.ipynb": ('python_basic', '.ipynb')
"filename": ('filename', '')
"/home/yungshun/workspace/py/python_w3resource_exercises/python_basic.ipynb": ('/home/yungshun/workspace/py/python_w3resource_exercises/python_basic', '.ipynb')
"/": ('/', '')
"": ('', '')


**107. Write a Python program to retrieve file properties.**

In [16]:
# `os.path` module & `time.ctime()` & `__file__`
import os.path, time

# Not work in Jupyter Notebook
# print("File:", __file__)
# print("Access time:", time.ctime(os.path.getatime(__file__)))
# print("Modified time:", time.ctime(os.path.getmtime(__file__)))
# print("Change time:", time.ctime(os.path.getctime(__file__)))
# print("Size:", os.path.getsize(__file__))

**108. Write a Python program to find the path to a file or directory when you encounter a path name.**

In [13]:
# [1] `os.path` module & `os.getcwd()`
import os.path

# Wrong solution in website
# Not work in Jupyter Notebook
# for file in [__file__, os.path.dirname(__file__), "/", "./broken_link"]:
#     print("File:", file)
#     print("Is Absolute?", os.path.isabs(file))
#     print("Is File?", os.path.isfile(file))
#     print("Is Dir?", os.path.isdir(file))
#     print("Is Link?", os.path.islink(file))
#     print("Exists?", os.path.exists(file))
#     print("Link Exists?", os.path.lexists(file))

# Not work in Jupyter Notebook
# print("Absolute file path:", os.path.abspath(__file__))

current_working_directory = os.getcwd()
print("Relative file path:", os.path.relpath("/home/yungshun/workspace/py/python-dsa", current_working_directory))

Relative file path: ../python-dsa


In [14]:
# [2] `os.path` module & `pathlib.Path.cwd()`
from pathlib import Path

# Not work in Jupyter Notebook
# print("Absolute file path:", os.path.abspath(__file__))

current_working_directory = Path.cwd()
print("Relative file path:", os.path.relpath("/home/yungshun/workspace/py/python-dsa", current_working_directory))

Relative file path: ../python-dsa


**109. Write a Python program to find the path to a file or directory when you encounter a path name.**

In [15]:
# [1] `os.path` module & `os.getcwd()`
import os.path

# Not work in Jupyter Notebook
# print("Absolute file path:", os.path.abspath(__file__))

current_working_directory = os.getcwd()
print("Relative file path:", os.path.relpath("/home/yungshun/workspace/py/python-dsa", current_working_directory))

Relative file path: ../python-dsa


In [16]:
# [2] `os.path` module & `pathlib.Path.cwd()`
from pathlib import Path

# Not work in Jupyter Notebook
# print("Absolute file path:", os.path.abspath(__file__))

current_working_directory = Path.cwd()
print("Relative file path:", os.path.relpath("/home/yungshun/workspace/py/python-dsa", current_working_directory))

Relative file path: ../python-dsa


**110. Write a Python program to get numbers divisible by fifteen from a list using an anonymous function.**

In [20]:
# [1] `filter()` & lambda expression
num_list = [15, 30, 45, 60, 70, 90]

# Use anonymous function to filter
print("Numbers of the said list divisible by 15 are", list(filter(lambda x: x % 15 == 0, num_list)))

Numbers divisible by 15 are [15, 30, 45, 60, 90]


In [24]:
# [2] Generator expression
num_list = [15, 30, 45, 60, 70, 90]

print("Numbers of the said list divisible by 15:", list([i for i in num_list if i % 15 == 0]))

Numbers of the said list divisible by 15: [15, 30, 45, 60, 90]


**111. Write a Python program to make file lists from the current directory using a wildcard.**

In [25]:
# [1] `glob` module
import glob

file_list = glob.glob("*.*")
print(file_list)

# Specific files
print(glob.glob("*.py"))
print(glob.glob("./[0-9].*"))

['python_basic_copy.py', 'python_basic.ipynb', 'tmp_array.npz', 'python_lambda.ipynb', 'python_data_types_list.ipynb', 'tmp_array.npy', 'python_basic.py', 'numpy_basic.ipynb']
['python_basic_copy.py', 'python_basic.py']
[]


In [27]:
# [2] `pathlib.Path()`
from pathlib import Path

for path in Path("/home/yungshun/workspace/py/python-w3resource-exercises").glob("*.*"):
   print(path)

/home/yungshun/workspace/py/python-w3resource-exercises/.ipynb_checkpoints
/home/yungshun/workspace/py/python-w3resource-exercises/python_basic_copy.py
/home/yungshun/workspace/py/python-w3resource-exercises/python_basic.ipynb
/home/yungshun/workspace/py/python-w3resource-exercises/.gitignore
/home/yungshun/workspace/py/python-w3resource-exercises/tmp_array.npz
/home/yungshun/workspace/py/python-w3resource-exercises/python_lambda.ipynb
/home/yungshun/workspace/py/python-w3resource-exercises/.git
/home/yungshun/workspace/py/python-w3resource-exercises/python_data_types_list.ipynb
/home/yungshun/workspace/py/python-w3resource-exercises/tmp_array.npy
/home/yungshun/workspace/py/python-w3resource-exercises/python_basic.py
/home/yungshun/workspace/py/python-w3resource-exercises/numpy_basic.ipynb


**112. Write a Python program to remove the first item from a specified list.**

In [30]:
# [1] List slicing
colors = ["Red", "Black", "Green", "White", "Orange"]
print("Original list elements:")
print(colors)

colors = colors[1:]
print("\nAfter removing the first element from the said list:")
print(colors)

Original list elements:
['Red', 'Black', 'Green', 'White', 'Orange']

After removing the first element from the said list:
['Black', 'Green', 'White', 'Orange']


In [31]:
# [2] `del`
colors = ["Red", "Black", "Green", "White", "Orange"]
print("Original list elements:")
print(colors)

del colors[0]
print("\nAfter removing the first element from the said list:")
print(colors)

Original list elements:
['Red', 'Black', 'Green', 'White', 'Orange']

After removing the first element from the said list:
['Black', 'Green', 'White', 'Orange']


In [33]:
# [3] `remove()`
colors = ["Red", "Black", "Green", "White", "Orange"]
print("Original list elements:")
print(colors)

# Remove first matching element
colors.remove("Red")
print("\nAfter removing the first element from the said list:")
print(colors)

Original list elements:
['Red', 'Black', 'Green', 'White', 'Orange']

After removing the first element from the said list:
['Black', 'Green', 'White', 'Orange']


In [35]:
# [4] Function & list slicing
def tail(lst):
    return lst[1:] if len(lst) > 1 else lst

print(tail([1, 2, 3, 4]))
print(tail([1]))
print(tail(["Red", "Black", "Green", "White", "Orange"]))

[2, 3, 4]
[1]
['Black', 'Green', 'White', 'Orange']


**113. Write a Python program that inputs a number and generates an error message if it is not a number.**

In [17]:
# [1] try-except block & `ValueError`
while True:
    try:
        num = float(input("Input a number: "))
        break
    except ValueError:
        print("This is not a number. Please try again.")

Input a number: python
This is not a number. Please try again.
Input a number: 16


In [34]:
# [2] `is_integer()` for checking integer (not number)
x = 3.14159
print("Check if {} is an integer:".format(x), x.is_integer())

Check if 3.14159 is an integer: False


In [36]:
# [3] `isinstance()` for checking integer (not number)
x = 3.14159
print("Check if {} is an integer:".format(x), isinstance(x, int))

Check if 3.14159 is an integer: False


**114. Write a Python program to filter positive numbers from a list.**

In [22]:
# [1] for loop
def filter_positive_numbers(num_list):
    filtered = []
    for i in num_list:
        if i > 0:
            filtered.append(i)
    return filtered

num_list = [-3, -2, -1, 0, 1, 2, 3]
print("Original numbers in the list:", num_list)
print("Positive numbers in the list:", filter_positive_numbers(num_list))

Original numbers in the list: [-3, -2, -1, 0, 1, 2, 3]
Positive numbers in the list: [1, 2, 3]


In [25]:
# [2] `filter()` & lambda expression
num_list = [-3, -2, -1, 0, 1, 2, 3]
print("Original numbers in the list:", num_list)

pos_num_list = list(filter(lambda n: n > 0, num_list))
print("Positive numbers in the list:", pos_num_list)

Original numbers in the list: [-3, -2, -1, 0, 1, 2, 3]
Positive numbers in the list: [1, 2, 3]


In [26]:
# [3] Generator expression
num_list = [-3, -2, -1, 0, 1, 2, 3]
print("Original numbers in the list:", num_list)

pos_num_list = [n for n in num_list if n > 0]
print("Positive numbers in the list:", pos_num_list)

Original numbers in the list: [-3, -2, -1, 0, 1, 2, 3]
Positive numbers in the list: [1, 2, 3]


**115. Write a Python program to compute the product of a list of integers (without using a for loop).**

In [29]:
# [1] `reduce()`
from functools import reduce

int_list = [10, 20, 30]
print(reduce(lambda x, y: x * y, int_list))

6000


In [30]:
# [2] `math.prod()`
import math

int_list = [10, 20, 30]
print(math.prod(int_list))

6000


**116. Write a Python program to print Unicode characters.**

In [31]:
# `u""`
string = u"\u0050\u0079\u0074\u0068\u006f\u006e \u0045\u0078\u0065\u0072\u0063\u0069\u0073\u0065\u0073 \u002d \u0077\u0033\u0072\u0065\u0073\u006f\u0075\u0072\u0063\u0065"
print(string)

Python Exercises - w3resource


**117. Write a Python program to prove that two string variables of the same value point to the same memory location.**

In [32]:
# `id()`
x = "python"
print("Memory location of x =", hex(id(x)))

y = "python"
print("Memory location of y =", hex(id(y)))

Memory location of x = 0x7f11f4c78470
Memory location of y = 0x7f11f4c78470


**118. Write a Python program to create a bytearray from a list.**

In [36]:
# `bytearray()`
num_list = [10, 12, 16, 36, 100]

values = bytearray(num_list)
print("Byte array:", values)

Byte array: bytearray(b'\n\x0c\x10$d')


**119. Write a Python program to round a floating-point number to a specified number of decimal places.**

In [60]:
# [1] `%`
num = 3.14159
print("Original floating point number: %f" % num)
print("Rounded floating point number: %.2f" % num)
print("Rounded floating point number: %.8f" % num)

Original floating point number: 3.141590
Rounded floating point number: 3.14
Rounded floating point number: 3.14159000


In [61]:
# [2] `format()`
num = 3.14159
print("Original floating point number: {:f}".format(num))
print("Rounded floating point number: {:0.2f}".format(num))
print("Rounded floating point number: {:0.8f}".format(num))

Original floating point number: 3.141590
Rounded floating point number: 3.14
Rounded floating point number: 3.14159000


In [74]:
# [3] f-strings
num = 3.14159
print(f"Original floating point number: {num}")
print(f"Rounded floating point number: {num:0.2f}")
print(f"Rounded floating point number: {num:0.8f}")

Original floating point number: 3.14159
Rounded floating point number: 3.14
Rounded floating point number: 3.14159000


**120. Write a Python program to format a specified string and limit the length of a string.**

In [70]:
# [1] `%`
string = "1234567890"
print("Orginal string: %s" % string)
print("Truncated string: %.2s" % string)
print("Truncated string: %.8s" % string)

Orginal string: 1234567890
Truncated string: 12
Truncated string: 12345678


In [72]:
# [2] `format()`
string = "1234567890"
print("Orginal string: {:s}".format(string))
print("Truncated string: {:.2s}".format(string))
print("Truncated string: {:.8s}".format(string))

Orginal string: 1234567890
Truncated string: 12
Truncated string: 12345678


In [73]:
# [3] f-strings
string = "1234567890"
print(f"Orginal string: {string}")
print(f"Truncated string: {string:.2s}")
print(f"Truncated string: {string:.8s}")

Orginal string: 1234567890
Truncated string: 12
Truncated string: 12345678


**121. Write a Python program to determine if a variable is defined or not.**

In [1]:
# `NameError`
try:
    x = 1
except NameError:
    print("Variable is not defined.")
else:
    print("Variable is defined.")

Variable is defined.


**122. Write a Python program to empty a variable without destroying it.**

<i>Sample data:</i> 

n = 20<br>
d = {"x": 200}

<i>Expected Output:</i> 

0<br>
{}

In [3]:
# [1] `type()()` & no function
n = 20
d = {"x": 200}
l = [1, 2, 3]
t = (1, 2, 3)

print(type(n)())
print(type(d)())
print(type(l)())
print(type(t)())

0
{}
[]
()


In [4]:
# [2] `type()()` & function
def empty_var(lst):
    return [type(i)() for i in lst]

lst = ["python", {"x": 12}, [10, 12, 16, "python"], (3, 36), 100]

print("Original objects:", lst)
print("Empty the variables without destroying them:", empty_var(lst))

Original objects: ['python', {'x': 12}, [10, 12, 16, 'python'], (3, 36), 100]
Empty the variables without destroying them: ['', {}, [], (), 0]


**123. Write a Python program to determine the largest and smallest integers, longs, and floats.**

In [6]:
# `sys` module
import sys

print("Float value information:", sys.float_info)
print("Integer value information:", sys.int_info)
print("Maximum size of an integer:", sys.maxsize)

Float value information: sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)
Integer value information: sys.int_info(bits_per_digit=30, sizeof_digit=4)
Maximum size of an integer: 9223372036854775807


**124. Write a Python program to check whether multiple variables have the same value.**

In [9]:
# [1] `==`
x = 12
y = 12
z = 12

if x == y == z == 12:
    print("All variables have same value.")

All variables have same value.


In [10]:
# [2] for loop
def multiple_variables_equality(*vars):
    for x in vars:
        if x != vars[0]:
            return "Variales have different values."
    return "All variables have same value."

print(multiple_variables_equality(1, 3, 1, 1, 1, 1))
print(multiple_variables_equality(16, 16, 16, 16))
print(multiple_variables_equality(-8, -8, -8, -8))

Variales have different values.
All variables have same value.
All variables have same value.


**125. Write a Python program to sum all counts in a collection.**

In [15]:
# `collections.Counter().values()` & `sum()`
import collections

nums = [6, 6, 8, 8, 16, 12, 100, 36, 36, 36]
print(sum(collections.Counter(nums).values()))

10


In [16]:
# `len()`
nums = [6, 6, 8, 8, 16, 12, 100, 36, 36, 36]
print(len(nums))

10


**126. Write a Python program to get the actual module object for a given object.**

In [17]:
# `inspect.getmodule()`
from inspect import getmodule
from math import sqrt

print(getmodule(sqrt))

<module 'math' from '/home/yungshun/anaconda3/lib/python3.8/lib-dynload/math.cpython-38-x86_64-linux-gnu.so'>


In [18]:
# `inspect.getmodule()` & function
import inspect

def add(x, y):
    return x + y

print(inspect.getmodule(add))

<module '__main__'>


**127. Write a Python program to check whether an integer fits in 64 bits.**

In [19]:
# `bit_length()`
int_val = 36
if int_val.bit_length() <= 63:
    print((-2 ** 63).bit_length())
    print((2 ** 63).bit_length())

64
64


**128. Write a Python program to check whether lowercase letters exist in a string.**

In [25]:
# [1] `any()` & `islower()` & no function
txt = "The quick brown fox jumps over the lazy dog."
print("Original string:", txt)
print("Lowercase letters exist in the string:", any(char.islower() for char in txt))

Original string: The quick brown fox jumps over the lazy dog.
Lowercase letters exist in the string: True


In [29]:
# [2] `any()` & `islower()` & function
def lowercase_string(txt):
    if any(char.islower() for char in txt):
        return True
    return False

txt = "The quick brown fox jumps over the lazy dog."
print("Original string:", txt)
print("Lowercase letters exist in the string:", lowercase_string(txt))

Original string: The quick brown fox jumps over the lazy dog.
Lowercase letters exist in the string: True


In [27]:
# [3] `ord()` & function
def lowercase_string(txt):
    ctr = 0
    for char in txt:
        if ord(char) >= 97 and ord(char) <= 122:
            ctr = ctr + 1
    if ctr > 0:
        return True
    
txt = "The quick brown fox jumps over the lazy dog."
print("Original string:", txt)
print("Lowercase letters exist in the string:", lowercase_string(txt))

Original string: The quick brown fox jumps over the lazy dog.
Lowercase letters exist in the string: True


**129. Write a Python program to add leading zeroes to a string.**

In [105]:
# [1] `zfill()`
string = "3.14159"
print("Original string:", string)
print("Added leading zeroes:", string.zfill(8))
print("Added leading zeroes:", string.zfill(16))

Original string: 3.14159
Added leading zeroes: 03.14159
Added leading zeroes: 0000000003.14159


In [106]:
# [2] f-string
string = "3.14159"
print("Original string:", string)
print("Added leading zeroes:", f"{string:>08}")

Original string: 3.14159
Added leading zeroes: 03.14159


In [107]:
# [3] `format()`
string = "3.14159"
string_format = "{:>08}"
print("Original string:", string)
print("Added leading zeroes:", string_format.format(string))

Original string: 3.14159
Added leading zeroes: 03.14159


In [108]:
# [4] `%`
string = 3.14159
float_string = float(string)
print("Original string:", num)
print("Added leading zeroes: %08.5f" % float_string)

Original string: 3.14159
Added leading zeroes: 03.14159


In [109]:
# [5] Trailing zeroes with `ljust()`
string = "3.14159"
print("Original string:", string)
print("Added trailing zeroes:", string.ljust(8, "0"))
print("Added trailing zeroes:", string.ljust(16, "0"))

Original string: 3.14159
Added trailing zeroes: 3.141590
Added trailing zeroes: 3.14159000000000


In [110]:
# [6] Trailing zeroes with f-string
string = "3.14159"
print("Original string:", string)
print("Added leading zeroe:", f"{string:<08}")

Original string: 3.14159
Added leading zeroe: 3.141590


In [111]:
# [7] Trailing zeroes with `format()`
string = "3.14159"
string_format = "{:<08}"
print("Original string:", string)
print("Added trailing zeroes:", string_format.format(string))

Original string: 3.14159
Added trailing zeroes: 3.141590


In [112]:
# [8] Trailing zeroes with `%`
string = 3.14159
float_string = float(string)
print("Original string:", string)
print("Added leading zeroes: %08f" % float_string)

Original string: 3.14159
Added leading zeroes: 3.141590


**130. Write a Python program that uses double quotes to display strings.**

In [115]:
# [1] f-string
string = "The quick brown fox jumps over the lazy dog."
print(f'"{string}"')

"The quick brown fox jumps over the lazy dog."


In [117]:
# [2] `format()`
string = "The quick brown fox jumps over the lazy dog."
print('"{}"'.format(string))

"The quick brown fox jumps over the lazy dog."


In [118]:
# [3] `%`
string = "The quick brown fox jumps over the lazy dog."
print('"%s"' % string)

"The quick brown fox jumps over the lazy dog."


In [119]:
# [4] `""`
string = '"The quick brown fox jumps over the lazy dog."'
print(string)

"The quick brown fox jumps over the lazy dog."


**131. Write a Python program to split a variable length string into variables.**

In [123]:
# Iterable unpacking
var_lst = ["a", "b", "c"]
x, y, z = var_lst
print(x, y, z)

a b c


**132. Write a Python program to list the home directory without an absolute path.**

In [125]:
# `os.path`
import os.path

print(os.path.expanduser("~"))

/home/yungshun


**133. Write a Python program to calculate the time runs (difference between start and current time) of a program.**

In [126]:
# [1] `timeit.default_timer()`
from timeit import default_timer

def timer(n):
    start = default_timer()
    for count in range(0, n):
        print(count)
    print(default_timer() - start)
    
timer(3)
timer(16)

0
1
2
4.267599433660507e-05
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
3.962393384426832e-05


In [131]:
# [2] `datetime.datetime.now()`
import datetime

def timer(n):
    start = datetime.datetime.now()
    for count in range(0, n):
        print(count)
    delta = datetime.datetime.now() - start
    print(delta)
    print(delta.days, "days")
    print(delta.seconds, "seconds")
    print(delta.microseconds, "microseconds")
    
timer(3)
timer(16)

0
1
2
0:00:00.000056
0 days
0 seconds
56 microseconds
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
0:00:00.000064
0 days
0 seconds
64 microseconds


In [132]:
# [3] `time.time()`
import time

def timer(n):
    start = time.time()
    for count in range(0, n):
        print(count)
    print(time.time() - start)

timer(3)
timer(16)

0
1
2
4.1961669921875e-05
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
2.5272369384765625e-05


In [134]:
# [4] `time.perf_counter()`
import time

def timer(n):
    start = time.perf_counter()
    for count in range(0, n):
        print(count)
    print(time.perf_counter() - start)

timer(3)
timer(16)

0
1
2
5.200703162699938e-05
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
6.25470420345664e-05


In [135]:
# [5] `time.process_time()`
import time

def timer(n):
    start = time.process_time()
    for count in range(0, n):
        print(count)
    print(time.process_time() - start)

timer(3)
timer(16)

0
1
2
6.767200000012963e-05
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
7.374599999998566e-05


**134. Write a Python program to input two integers on a single line.**

**135. Write a Python program to print a variable without spaces between values.**

<i>Sample value:</i> 

x =30

<i>Expected output:</i> 

Value of x is "30"